lib/instrumentation/modules/aws-sdk/sqs.js

```diff
@@ -0,0 +1,193 @@
+      type: TYPE
+    },
+    cloud: {
+      region: getRegionFromRequest(request)
+    }
+  }
+  return destination
+}
+
+// Record queue related metrics
+//
+// Creates metric collector objects on first run, and
+// updates their data with data from received messages
+function recordMetrics (queueName, data, agent) {
+  const messages = data && data.Messages
+  if (!messages || messages.length < 1) {
+    return
+  }
+  if (!queueMetics.get(queueName)) {
+    const collector = agent._metrics.createQueueMetricsCollector(queueName)
+    queueMetics.set(queueName, collector)
+  }
+  const metrics = queueMetics.get(queueName)
+
+  for (const message of messages) {
+    const sentTimestamp = message.Attributes && message.Attributes.SentTimestamp
+    const delay = (new Date()).getTime() - sentTimestamp
+    metrics.updateStats(delay)
+  }
+}
+
+// Creates the span name from request information
+function getSpanNameFromRequest (request) {
+  const action = getActionFromRequest(request)
+  const toFrom = getToFromFromOperation(request.operation)
+  const queueName = getQueueNameFromRequest(request)
+
+  const name = `${SUBTYPE.toUpperCase()} ${action.toUpperCase()} ${toFrom} ${queueName}`
+  return name
+}
+
+function shouldIgnoreRequest (request, agent) {
+  const operation = request && request.operation
+  // are we interested in this operation/method call?
+  if (OPERATIONS.indexOf(operation) === -1) {
+    return true
+  }
+
+  // is the named queue on our ignore list?
+  if (agent._conf && agent._conf.ignoreMessageQueuesRegExp) {
+    const queueName = getQueueNameFromRequest(request)
+    for (const rule of agent._conf.ignoreMessageQueuesRegExp) {
+      if (rule.test(queueName)) {
+        return true
+      }
+    }
+  }
+
+  if (!agent.currentTransaction) {
+    agent.logger.trace('no active transaction found, skipping sqs instrumentation')
+    return true
+  }
+
+  return false
+}
+
+function wrapReceiveMessageCallback (original, request, agent) {
+  if (typeof original !== 'function' || original.name === 'elasticReceiveMessageCallback') return original
+
+  return elasticReceiveMessageCallback
+  function elasticReceiveMessageCallback (err, data) {
+    recordMetrics(getQueueNameFromRequest(request), data, agent)
+    return original.apply(this, [err, data])
+  }
+}
+
+// Main entrypoint for SQS instrumentation
+//
+// Must call (or one of its function calls must call) the
+// `orig` function/method
+function instrumentationSqs (orig, origArguments, request, AWS, agent, { version, enabled }) {
+  if (shouldIgnoreRequest(request, agent)) {
+    return orig.apply(request, origArguments)
+  }
+
+  const type = TYPE
+  const subtype = SUBTYPE
+  const action = getActionFromRequest(request)
+  const name = getSpanNameFromRequest(request)
+  const span = agent.startSpan(name, type, subtype, action)
+  span.setDestinationContext(getMessagingDestinationContextFromRequest(request))
```

Comment:
Should we also be added `span.context.message.*` here, per https://github.com/elastic/apm/blob/master/specs/agents/tracing-instrumentation-messaging.md#context-fields
`message.age.ms` and `message.queue.name`.

Reply:
Good catch.  We can support queue.name for sure -- but the message age will be less straight forward.I've opened an issue for future consideration here: https://github.com/elastic/apm-agent-nodejs/issues/2026